### ISSUE TITLE
supporting operations transparently across multiple GPUs

### ISSUE THREAD

**Issue by gonzaloiglesiasiglesias** [arrayfire/arrayfire#2463]
It seems that AF cannot run operations transparantely across multiple gpus. For instance, I'm trying to do the following:
 
std::vector<float> d = {1};
  af::setDevice(0);
  const af::array l1(1, d.data());
  af::setDevice(1);
  const af::array l2(1, d.data());
  af::array r = l1 + l2;
  std::vector<float> o(1);
  r.host(&o[0]);

This yields the error:
 AfError: Input Array not created on current device

It would be really cool feature if arrayfire could handle this by transparently. 

Related:
- given the af::array l1, is there a method or function to know to retrieve that it was created in device 0, or do i have to do the book keeping myself?

**Comment by 9prady9** [comment]
Yes, that would be nice feature to have, no doubt about that.

Making that feature available across all backends would be however a tantamount task.
- CPU backend would be the easiest I think
- CUDA can be easier than OpenCL with peer access enabled. However, some design decisions will be tricky like a couple of them given below
    - should a copy(`cudaMemcpyPeerAsync`) of original(source device memory) data be made on target device for faster access? If so, under what conditions...
    - should peer access be automatically enabled across all devices or should this be user choice 
- OpenCL: Some kind of support for peer to peer direct access is made available in [ROCm](https://rocm.github.io/ROCmMultiGPU.html), but I haven't used and not aware of what features are available. In the worst case scenario, copy via CPU is the way to have the feature but I doubt we should make that choice for users. The user has to choose what kind of solution they would be employing in their use case. There is DirectGMA but only for FirePro GPUs, it doesn't seem to be like a universal solution for all AMD cards.

**Comment by WilliamTambellini** [comment]
@9prady9 
Would NCCL help : 
https://developer.nvidia.com/nccl
for this simple use case: 
array A on gpu 1
array B on gpu 2
array C on gpu 3
C = A + B; // no exception, AF smartly takes care of cross gpu data transfer
?
Cheers
W.

**Comment by 9prady9** [comment]
Yes, It probably will help eliminate the responsibility of efficient data transfers. However I have some concerns/questions based on what I gathered from NCCL [documentation](https://docs.nvidia.com/deeplearning/sdk/nccl-developer-guide/docs/index.html).

- Should ArrayFire target single-node multi-GPU setups or use NCCL for multi-node(networked) setups. What course should OpenCL backend take if we take NCCL route for CUDA backend to preserve feature parity ?
- For single node configuration
    - How much of overhead NCCL would be if ArrayFire handles single-node multi-GPU configuration only
    - Would CUDA-peer-memory-access with some event handling would be more efficient than NCCL? Again, we would have to do some benchmarking to confirm this.
- It looks like [MacOS](https://github.com/NVIDIA/nccl/issues/201) and [Windows](https://github.com/NVIDIA/nccl/pull/31) support for NCCL isn't ready yet.

**Comment by WilliamTambellini** [comment]
Hi @9prady9 
Thanks. Good questions.
I would consequently be in favor of a minimalist 1st implementation (without NCCL) : 
1- if the computation tree to be evaluated mixes up arrays from different devices, AF to internally centralize/transfer the arrays to a single GPU (the one with maximum free space ?)
2- AF to eval/exec the tree using only this "central" gpu 

Would that makes sens ?
Cheers
W.

**Comment by 9prady9** [comment]
I personally would prefer not adding one more dependency to ArrayFire. As far as peer-to-peer transfers are considered, I am not sure what approach is the best. May be @umar456 has some thoughts on it.
